infrastructure/oss/types.go

```diff
@@ -243,6 +290,68 @@ func (i *ossIssue) AddSnykLearnAction(learnService learn.Service, ep error_repor
 	return action
 }
 
+func (i *ossIssue) AddQuickFixAction(affectedFilePath string, issueRange snyk.Range) *snyk.CodeAction {
+	if !config.CurrentConfig().IsSnyOSSQuickFixCodeActionsEnabled() {
+		return nil
+	}
+	log.Debug().Msg("create deferred quickfix code action")
+	quickfixEdit := i.getQuickfixEdit(affectedFilePath)
+	if quickfixEdit == "" {
+		return nil
+	}
+	upgradeMessage := "Upgrade to " + quickfixEdit + " (Snyk)"
+	autofixEditCallback := func() *snyk.WorkspaceEdit {
+		edit := &snyk.WorkspaceEdit{}
+		singleTextEdit := snyk.TextEdit{
+			Range:   issueRange,
+			NewText: quickfixEdit,
+		}
+		edit.Changes = make(map[string][]snyk.TextEdit)
+		edit.Changes[affectedFilePath] = []snyk.TextEdit{singleTextEdit}
+		return edit
+	}
+
+	action, err := snyk.NewDeferredCodeAction(upgradeMessage, &autofixEditCallback, nil)
+	if err != nil {
+		log.Error().Msg("failed to create deferred quickfix code action")
+		return nil
+	}
+	return &action
+}
+
+func (i *ossIssue) getQuickfixEdit(affectedFilePath string) string {
+	hasUpgradePath := len(i.UpgradePath) > 1
+	if !hasUpgradePath {
+		return ""
+	}
+
+	// UpgradePath[0] is the upgrade for the package that was scanned
+	// UpgradePath[1] is the upgrade for the root dependency
+	rootDependencyUpgrade := strings.Split(i.UpgradePath[1].(string), "@")
+	depName := strings.Join(rootDependencyUpgrade[:len(rootDependencyUpgrade)-1], "@")
+	depVersion := rootDependencyUpgrade[len(rootDependencyUpgrade)-1]
+	if i.PackageManager == "npm" || i.PackageManager == "yarn" || i.PackageManager == "yarn-workspace" {
+		return fmt.Sprintf("\"%s\": \"%s\"", depName, depVersion)
+	} else if i.PackageManager == "maven" {
+		depNameSplit := strings.Split(depName, ":")
+		depName = depNameSplit[len(depNameSplit)-1]
+		// TODO: remove once https://snyksec.atlassian.net/browse/OSM-1775 is fixed
+		if strings.Contains(affectedFilePath, "build.gradle") {
+			return fmt.Sprintf("%s:%s", depName, depVersion)
+		}
+		return depVersion
+	} else if i.PackageManager == "gradle" {
+		depNameSplit := strings.Split(depName, ":")
+		depName = depNameSplit[len(depNameSplit)-1]
+		return fmt.Sprintf("%s:%s", depName, depVersion)
+	}
+	if i.PackageManager == "gomodules" {
+		return fmt.Sprintf("v%s", depVersion)
+	}
+
```

Comment:
Idea: we could have a default handler, that handles other package managers (e.g. requirements.txt of python, php etc). We'd need to test it though with these package managers. Also, would most likely be a follow-up story.

Reply:
I've written up https://snyksec.atlassian.net/browse/IDE-473 with what I think you're suggesting, feel free to edit it!